docs/specs/instrumentation.md

```diff
@@ -8,34 +8,145 @@ interpreted as described in [RFC 2119].
+    example, a failed HTTP request that will be retried.
+  * `error_code` - An error code for the failure, if applicable.
+    * SHOULD only be specified if it adds additional information beyond
+      `error_type`.
+    * The values for `error_code` for a given error event MUST be a bounded set
+      with relatively low cardinality because it will be used as a metric tag.
+      Examples would be syscall error code. Examples of values that should not
+      be used are raw error messages from `serde` as these are highly variable
+      depending on the input. Instead, these errors should be converted to an
+      error code like `invalid_json`.
+  * `error_type` - The type of error condition. MUST be one of the types listed
+    in the `error_type` enum list in the cue docs.
+  * `stage` - The stage at which the error occurred. MUST be one of `receiving`,
+    `processing`, or `sending`.
+  * If any of the above properties are implicit to the specific error
+    type, they MAY be omitted from being represented explicitly in the
+    event fields. However, they MUST still be included in the emitted
+    logs and metrics, as specified below, as if they were present.
+* Metrics
+  * MUST include the defined properties as tags.
+  * If `retriable` is `true`, MUST increment `<namespace>_warnings_total` metric.
+  * If `retriable` is `false`, MUST increment `<namespace>_errors_total` metric.
+* Logs
+  * MUST log a descriptive, user friendly error message that sufficiently
+    describes the error.
+  * MUST include the defined properties as key-value pairs, except `message`.
+  * If `retriable` is `true`, MUST log a message at the `warning` level.
+  * If `retriable` is `false`, MUST log a message at the `error` level.
+  * SHOULD be rate limited to 10 seconds.
+* Events
+  * MUST emit an [`EventsDropped`] event if the error results in dropping events,
+    or the error itself MUST meeting the `EventsDropped` requirements.
+
+#### EventsDropped
+
+An `<Namespace>EventsDropped` event must be emitted when events are dropped.
+If events are dropped due to an error, then the error event should drive the
+emission of this event, meeting the following requirements:
+
+* Properties
+  * `intentional` - Distinguishes if the events were dropped intentionally. For
+    example, events dropped in the `filter` transform are intentionally dropped,
+    while events dropped due to an error in the `remap` transform are
+    unintentionally dropped.
+* Metrics
+  * MUST increment the `<namespace>_discarded_events_total` counter by the
+    number of events discarded.
+  * MUST NOT increment this metric if retrying the operation will preserve the
+    event, such as retrying delivery in sinks.
+  * MUST include the `intentional` property as a tag.
+* Logs
+  * MUST log a `<count> events [un]intentionally dropped due to <reason>`
+    message. `<reason>` MUST be a descriptive, user-friendly message that helps
+    the user understand why their data was dropped.
```

Comment:
I think `<reason>` and `<count>` should be attributes of the log message rather than being within the string message.

Reply:
Resolved in https://github.com/vectordotdev/vector/pull/12822/commits/5593513d62c9dab682eecaad96f30a9cbad9689b